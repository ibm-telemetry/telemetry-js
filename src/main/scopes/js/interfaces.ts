/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type * as ts from 'typescript'

import { Logger } from '../../core/log/logger.js'
import { ComplexValue } from './complex-value.js'
import { JsNodeHandler } from './node-handlers/js-node-handler.js'
import { NodeValueHandler } from './node-handlers/value-handlers/node-value-handler.js'

export interface JsImport {
  name: string
  path: string
  isDefault: boolean
  isAll: boolean
  rename?: string
}

type JsNodeHandlerClass<DataType> = new (
  node: ts.SourceFile,
  logger: Logger
) => JsNodeHandler<DataType>

export type JsNodeHandlerMap = Partial<Record<ts.SyntaxKind, JsNodeHandlerClass<unknown>>>

export type NodeValue = string | number | boolean | ComplexValue | null | undefined

type NodeValueHandlerProducer = new (node: ts.SourceFile, logger: Logger) => NodeValueHandler

export type NodeValueHandlerMap = Partial<Record<ts.SyntaxKind, NodeValueHandlerProducer>>
