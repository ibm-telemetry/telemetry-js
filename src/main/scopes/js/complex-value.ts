/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Object representing a complex value.
 */
export class ComplexValue {
  constructor(public readonly complexValue: unknown) {}
}
