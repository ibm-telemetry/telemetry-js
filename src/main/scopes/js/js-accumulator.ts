/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type { JsImport } from '../js/interfaces.js'

/**
 * Base class for all JS Accumulators.
 * Responsible for maintaining an aggregated state of imports and other elements.
 */
export abstract class JsAccumulator {
  public imports: JsImport[]

  constructor() {
    this.imports = []
  }
}
