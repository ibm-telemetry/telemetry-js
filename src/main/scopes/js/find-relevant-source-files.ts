/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import getPropertyByPath = require('lodash/get.js')
import * as path from 'node:path'

import { ObjectPath } from 'object-scan'

import { getTrackedSourceFiles } from '../../core/get-tracked-source-files.js'
import { Logger } from '../../core/log/logger.js'
import { NoInstallationFoundError } from '../../exceptions/no-installation-found-error.js'
import { getDependencyTree } from '../npm/get-dependency-tree.js'
import { getDirectoryPrefix } from '../npm/get-directory-prefix.js'
import { getInstalledVersionPaths } from '../npm/get-installed-version-paths.js'
import { getPackageData } from '../npm/get-package-data.js'
import { getPackageTrees } from '../npm/get-package-trees.js'
import { getTreePredecessor } from '../npm/get-tree-predecessor.js'
import { DependencyTree, PackageData } from '../npm/interfaces.js'

/**
 * Finds tracked source files and then filters them based on ones that appear in a  project which
 * depends on the in-context instrumented package/version.
 *
 * @param instrumentedPackage - Data about the instrumented package to use during filtering.
 * @param cwd - Current working directory. This must be inside of the root directory. This is an
 * absolute path.
 * @param root - Root-most directory. This is an absolute path.
 * @param fileExtensions - List of file extensions to capture metrics for.
 * @param logger - Logger instance.
 * @returns A (possibly empty) array of source files.
 */
export async function findRelevantSourceFiles(
  instrumentedPackage: PackageData,
  cwd: string,
  root: string,
  fileExtensions: string[],
  logger: Logger
) {
  logger.traceEnter('', 'findRelevantSourceFiles', [instrumentedPackage, cwd, root, fileExtensions])
  const sourceFiles = await getTrackedSourceFiles(root, logger, fileExtensions)

  const dependencyTree = await getDependencyTree(cwd, root, logger)

  const filterPromises = sourceFiles.map(async (f) => {
    const prefix = await getDirectoryPrefix(path.dirname(f.fileName), logger)
    const prefixPackageData = await getPackageData(prefix, root, logger)

    let packageTrees = getPackageTrees(dependencyTree, prefixPackageData)

    let instrumentedInstallVersions: string[] | undefined = undefined
    let shortestPathLength: number | undefined = undefined
    do {
      for (const tree of packageTrees) {
        const instrumentedInstallPaths = getInstalledVersionPaths(tree, instrumentedPackage.name)
        if (instrumentedInstallPaths.length > 0) {
          const pathsLength = instrumentedInstallPaths[0]?.length ?? 0
          if (shortestPathLength === undefined || pathsLength < shortestPathLength) {
            instrumentedInstallVersions = instrumentedInstallPaths.map(
              (path) => getPropertyByPath(tree, path)['version']
            )
            shortestPathLength = pathsLength
          }
        }
      }
      // did not find, go up one level for all packages
      packageTrees = packageTrees
        .map((tree) => getTreePredecessor(dependencyTree, tree['path'] as ObjectPath))
        .filter((tree) => tree !== undefined) as DependencyTree[]
    } while (shortestPathLength === undefined && packageTrees.length > 0)

    if (instrumentedInstallVersions === undefined) {
      throw new NoInstallationFoundError(instrumentedPackage.name)
    }

    return instrumentedInstallVersions.some((version) => version === instrumentedPackage.version)
  })
  const filterData = await Promise.all(filterPromises)

  const results = sourceFiles.filter((_, index) => {
    return filterData[index]
  })

  logger.traceEnter('', 'findRelevantSourceFiles', results)
  return results
}
