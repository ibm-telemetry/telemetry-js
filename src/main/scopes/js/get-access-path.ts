/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as ts from 'typescript'

import type { Logger } from '../../core/log/logger.js'
import { ComplexValue } from './complex-value.js'
import { getNodeValueHandler } from './node-value-handler-map.js'

/**
 * Constructs a JsToken object from a given Identifier type AST node.
 *
 * @param node - TS node to retrieve access path for.
 * @param sourceFile - Root AST node.
 * @param logger - A logger instance.
 * @param currAccessPath - For internal use only, tracks current constructed access path.
 * @param topLevel - For internal use only, tracks top level function call.
 * @returns Access path to node represented as string array (chunks).
 */
export default function getAccessPath(
  node: ts.Node,
  sourceFile: ts.SourceFile,
  logger: Logger,
  currAccessPath: string[] = [],
  topLevel: boolean = true
): string[] {
  switch (node.kind) {
    case ts.SyntaxKind.Identifier:
      return [...currAccessPath, (node as ts.Identifier).escapedText.toString()]
    case ts.SyntaxKind.PropertyAccessExpression:
      currAccessPath.push((node as ts.PropertyAccessExpression).name.escapedText.toString())
      break
    case ts.SyntaxKind.ElementAccessExpression: {
      const argumentExpression = (node as ts.ElementAccessExpression).argumentExpression
      const data = getNodeValueHandler(argumentExpression.kind, sourceFile, logger).getData(
        argumentExpression
      )
      if (data instanceof ComplexValue) {
        currAccessPath = [
          ...currAccessPath,
          getAccessPath(argumentExpression, sourceFile, logger, [], false).reverse().join('.')
        ]
      } else {
        currAccessPath.push(data?.toString() ?? '')
      }
      break
    }
  }

  let accessPath = currAccessPath

  if ('expression' in node) {
    accessPath = getAccessPath(
      node.expression as ts.Node,
      sourceFile,
      logger,
      currAccessPath,
      false
    )
  }

  return topLevel ? accessPath.reverse() : accessPath
}
