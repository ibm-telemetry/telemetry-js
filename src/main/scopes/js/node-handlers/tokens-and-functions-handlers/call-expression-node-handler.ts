/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type * as ts from 'typescript'

import getAccessPath from '../../get-access-path.js'
import type { JsFunction } from '../../interfaces.js'
import type { JsFunctionTokenAccumulator } from '../../js-function-token-accumulator.js'
import { getNodeValueHandler } from '../../node-value-handler-map.js'
import { JsNodeHandler } from '../js-node-handler.js'

/**
 * Holds logic to construct a JsFunction object given an CallExpression node.
 *
 */
export class CallExpressionNodeHandler extends JsNodeHandler<JsFunction> {
  /**
   * Processes a CallExpression node data and adds it to the given accumulator.
   *
   * @param node - Node element to process.
   * @param accumulator - JsFunctionTokenAccumulator instance
   * that holds the aggregated functions state.
   */
  handle(node: ts.CallExpression, accumulator: JsFunctionTokenAccumulator) {
    const jsFunction = this.getData(node)

    accumulator.functions.push(jsFunction)
  }

  /**
   * Constructs a JsFunction object from a given CallExpression type AST node.
   *
   * @param node - Node element to process.
   * @returns Constructed JsFunction object.
   */
  getData(node: ts.CallExpression): JsFunction {
    const argsLength = node.arguments.end - node.arguments.pos
    const nodeText = node.getText(this.sourceFile)
    const jsFunction: JsFunction = {
      // account for parentheses
      name: nodeText.substring(0, nodeText.length - argsLength - 2),
      accessPath: [],
      arguments: [],
      startPos: node.pos,
      endPos: node.end
    }

    jsFunction.arguments = node.arguments.map((arg) =>
      getNodeValueHandler(arg.kind, this.sourceFile, this.logger).getData(arg)
    )

    jsFunction.accessPath = getAccessPath(node, this.sourceFile, this.logger)

    return jsFunction
  }
}
