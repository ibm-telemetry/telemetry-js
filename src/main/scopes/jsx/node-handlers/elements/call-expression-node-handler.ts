/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as ts from 'typescript'

import { DEFAULT_ELEMENT_NAME } from '../../constants.js'
import { JsxImport } from '../../interfaces.js'
import { type JsxElementAccumulator } from '../../jsx-element-accumulator.js'
import { ElementNodeHandler } from './element-node-handler.js'

/**
 * Holds logic to construct a CallExpression object given a node of said kind.
 *
 */
export class CallExpressionNodeHandler extends ElementNodeHandler<JsxImport[]> {
  /**
   * Processes a CallExpression node data and adds it to the given accumulator.
   *
   * @param node - Node element to process.
   * @param accumulator - JsxAccumulator instance that holds the aggregated elements state.
   */
  handle(node: ts.CallExpression, accumulator: JsxElementAccumulator) {
    accumulator.imports.push(...this.getData(node))
  }

  /**
   * Constructs a JsxImport object from a given CallExpression type AST node.
   *
   * @param node - Node element to process.
   * @returns Constructed JsxElement object.
   */
  getData(node: ts.CallExpression): JsxImport[] {
    const imports: JsxImport[] = []
    let path: string | undefined
    if (ts.isPropertyAccessExpression(node.expression)) {
      if (ts.isCallExpression(node.expression.expression)) {
        if (node.expression.expression.expression.kind === ts.SyntaxKind.ImportKeyword) {
          path = node.expression.expression.arguments[0]?.getText(this.sourceFile)
        }
      }
    }
    if (path === undefined) return []
    if (node.arguments[0] && ts.isArrowFunction(node.arguments[0])) {
      const parameter = node.arguments[0].parameters[0]

      if (parameter !== undefined && ts.isIdentifier(parameter.name)) {
        imports.push({
          name: DEFAULT_ELEMENT_NAME,
          rename: parameter.name.getText(this.sourceFile),
          path,
          isDefault: true,
          isAll: false
        })
      }
      if (parameter !== undefined && ts.isObjectBindingPattern(parameter.name)) {
        parameter.name.elements.forEach((element) => {
          imports.push({
            name: element.name.getText(this.sourceFile),
            path: path as string,
            isDefault: false,
            isAll: false
          })
        })
      }
    }

    return imports
  }
}
