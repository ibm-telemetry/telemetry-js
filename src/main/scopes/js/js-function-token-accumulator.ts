/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { JsFunction, JsImport, JsToken } from '../js/interfaces.js'
import { JsAccumulator } from '../js/js-accumulator.js'

/**
 * Responsible for maintaining an aggregated state of imports, functions and tokens.
 */
export class JsFunctionTokenAccumulator extends JsAccumulator {
  public readonly tokens: JsToken[]
  public readonly functions: JsFunction[]
  public readonly functionImports: Map<JsFunction, JsImport>
  public readonly tokenImports: Map<JsToken, JsImport>

  constructor() {
    super()
    this.tokens = []
    this.functions = []
    this.functionImports = new Map()
    this.tokenImports = new Map()
  }
}
