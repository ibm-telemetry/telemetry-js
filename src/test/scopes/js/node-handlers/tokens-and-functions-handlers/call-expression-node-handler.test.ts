/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as ts from 'typescript'
import { describe, expect, it } from 'vitest'

import { ComplexValue } from '../../../../../main/scopes/js/complex-value.js'
import { JsFunctionTokenAccumulator } from '../../../../../main/scopes/js/js-function-token-accumulator.js'
// eslint-disable-next-line max-len -- It's a long import
import { CallExpressionNodeHandler } from '../../../../../main/scopes/js/node-handlers/tokens-and-functions-handlers/call-expression-node-handler.js'
import { createSourceFileFromText } from '../../../../__utils/create-source-file-from-text.js'
import { findNodesByType } from '../../../../__utils/find-nodes-by-type.js'
import { initLogger } from '../../../../__utils/init-logger.js'

describe('class: CallExpressionExpressionNodeHandler', async () => {
  const logger = initLogger()

  it('captures function at the end of a property chain', () => {
    const accumulator = new JsFunctionTokenAccumulator()
    const sourceFile = createSourceFileFromText('foo.bar.baz()')
    const nodes = findNodesByType<ts.CallExpression>(sourceFile, ts.SyntaxKind.CallExpression)
    const handler = new CallExpressionNodeHandler(sourceFile, logger)

    nodes.forEach((node) => {
      handler.handle(node, accumulator)
    })

    expect(accumulator.functions).toHaveLength(1)
    expect(accumulator.functions[0]).toStrictEqual({
      name: 'foo.bar.baz',
      arguments: [],
      accessPath: ['foo', 'bar', 'baz'],
      startPos: 0,
      endPos: 13
    })
  })

  it('captures function at the end of a property chain', () => {
    const accumulator = new JsFunctionTokenAccumulator()
    const sourceFile = createSourceFileFromText('foo[ACCESS_TOKEN["bla"]].baz()')
    const nodes = findNodesByType<ts.CallExpression>(sourceFile, ts.SyntaxKind.CallExpression)
    const handler = new CallExpressionNodeHandler(sourceFile, logger)

    nodes.forEach((node) => {
      handler.handle(node, accumulator)
    })

    expect(accumulator.functions).toHaveLength(1)
    expect(accumulator.functions[0]).toMatchObject({
      name: 'foo[ACCESS_TOKEN["bla"]].baz',
      arguments: [],
      accessPath: ['foo', new ComplexValue('ACCESS_TOKEN["bla"]'), 'baz']
    })
  })

  it('captures a function for a intermediate function', () => {
    const accumulator = new JsFunctionTokenAccumulator()
    const sourceFile = createSourceFileFromText('foo.bar().baz')
    const nodes = findNodesByType<ts.CallExpression>(sourceFile, ts.SyntaxKind.CallExpression)
    const handler = new CallExpressionNodeHandler(sourceFile, logger)

    nodes.forEach((node) => {
      handler.handle(node, accumulator)
    })

    expect(accumulator.functions).toHaveLength(1)
    expect(accumulator.functions[0]).toMatchObject({
      name: 'foo.bar',
      arguments: [],
      accessPath: ['foo', 'bar']
    })
  })

  it('captures a function for a simple function', () => {
    const accumulator = new JsFunctionTokenAccumulator()
    const sourceFile = createSourceFileFromText('foo()')
    const nodes = findNodesByType<ts.CallExpression>(sourceFile, ts.SyntaxKind.CallExpression)
    const handler = new CallExpressionNodeHandler(sourceFile, logger)

    nodes.forEach((node) => {
      handler.handle(node, accumulator)
    })

    expect(accumulator.functions).toHaveLength(1)
    expect(accumulator.functions[0]).toMatchObject({
      name: 'foo',
      arguments: [],
      accessPath: ['foo']
    })
  })

  it('??', () => {
    const accumulator = new JsFunctionTokenAccumulator()
    const sourceFile = createSourceFileFromText('foo().faa()')
    const nodes = findNodesByType<ts.CallExpression>(sourceFile, ts.SyntaxKind.CallExpression)
    const handler = new CallExpressionNodeHandler(sourceFile, logger)

    nodes.forEach((node) => {
      handler.handle(node, accumulator)
    })

    expect(accumulator.functions).toHaveLength(2)
    expect(accumulator.functions[1]).toMatchObject({
      name: 'foo',
      arguments: [],
      accessPath: ['foo']
    })
    expect(accumulator.functions[0]).toMatchObject({
      name: 'foo().faa',
      arguments: [],
      accessPath: ['foo', 'faa']
    })
  })

  describe('functions containing arguments', () => {
    it('captures a function and its arguments', () => {
      const accumulator = new JsFunctionTokenAccumulator()
      const sourceFile = createSourceFileFromText('foo(first)')
      const nodes = findNodesByType<ts.CallExpression>(sourceFile, ts.SyntaxKind.CallExpression)
      const handler = new CallExpressionNodeHandler(sourceFile, logger)

      nodes.forEach((node) => {
        handler.handle(node, accumulator)
      })

      expect(accumulator.functions).toHaveLength(1)
      expect(accumulator.functions[0]).toMatchObject({
        name: 'foo',
        arguments: [new ComplexValue('first')],
        accessPath: ['foo']
      })
    })
  })
})
