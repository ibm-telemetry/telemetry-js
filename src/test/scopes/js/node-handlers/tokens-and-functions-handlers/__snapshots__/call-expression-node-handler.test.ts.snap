// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`class: CallExpressionNodeHandler > correctly returns the JsFunctions for a complex fixture 1`] = `
[
  {
    "accessPath": [
      "someFunction",
    ],
    "arguments": [],
    "endPos": 207,
    "name": "someFunction",
    "startPos": 0,
  },
  {
    "accessPath": [
      "anObject",
      "property",
      "optionallyChainedFunction",
    ],
    "arguments": [
      1,
      ComplexValue {
        "complexValue": "[2,3,4]",
      },
    ],
    "endPos": 267,
    "name": "optionallyChainedFunction",
    "startPos": 207,
  },
  {
    "accessPath": [
      "anObject",
      "property",
      "aFunction",
    ],
    "arguments": [
      "aNormalArgument",
    ],
    "endPos": 315,
    "name": "aFunction",
    "startPos": 267,
  },
  {
    "accessPath": [
      "anObject",
      "property",
      "aFunction",
    ],
    "arguments": [],
    "endPos": 349,
    "name": "aFunction",
    "startPos": 315,
  },
  {
    "accessPath": [
      "anObject",
      "property",
      "4",
    ],
    "arguments": [
      "hey",
      500,
      ComplexValue {
        "complexValue": "{yeah: 'yeah'}",
      },
    ],
    "endPos": 402,
    "name": "4",
    "startPos": 349,
  },
  {
    "accessPath": [
      "anObject",
      "BLA",
      "property",
      "aFunction",
    ],
    "arguments": [
      ComplexValue {
        "complexValue": "{variable, ...object}",
      },
    ],
    "endPos": 465,
    "name": "'aFunction'",
    "startPos": 402,
  },
  {
    "accessPath": [
      "anObject",
      "property",
      "aFunction",
    ],
    "arguments": [
      "click",
      ComplexValue {
        "complexValue": "aFunction",
      },
    ],
    "endPos": 516,
    "name": "aFunction",
    "startPos": 465,
  },
  {
    "accessPath": [
      "aFunction",
      "anotherFunction",
      "anotherFunction",
    ],
    "arguments": [],
    "endPos": 593,
    "name": "anotherFunction",
    "startPos": 516,
  },
]
`;
